apps/marginfi-v2-trading/src/pages/api/oracle/price.ts

```diff
@@ -154,16 +200,115 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+      return crossbarPayloadToOraclePricePerFeedHash(payload);
+    }
+
+    if (process.env.SWITCHBOARD_CROSSSBAR_API_FALLBACK) {
+      // fallback crossbar
+      const { payload: fallbackPayload, brokenFeeds: fallbackBrokenFeeds } = await fetchCrossbarPrices(
+        mainBrokenFeeds,
+        process.env.SWITCHBOARD_CROSSSBAR_API_FALLBACK
+      );
+      payload.push(...fallbackPayload);
+
+      if (!fallbackBrokenFeeds.length) {
+        return crossbarPayloadToOraclePricePerFeedHash(payload);
+      }
+    }
+
+    // birdeye as last resort
+    const { payload: birdeyePayload, brokenFeeds: birdeyeBrokenFeeds } = await fetchBirdeyePrices(feedHashes, mintMap);
+
+    payload.push(...birdeyePayload);
+
+    birdeyeBrokenFeeds.forEach((feed) => {
+      payload.push({
+        feedHash: feed,
+        results: [0],
+      });
+    });
+
+    return crossbarPayloadToOraclePricePerFeedHash(payload);
+  } catch (error) {
+    console.error("Error:", error);
+    throw new Error("Couldn't fetch from crossbar");
+  }
+}
+
+async function fetchBirdeyePrices(
+  feedHashes: string[],
+  mintMap: Map<string, PublicKey>
+): Promise<{ payload: CrossbarSimulatePayload; brokenFeeds: string[] }> {
+  try {
+    const brokenFeeds: string[] = [];
+
+    const tokens = feedHashes
+      .map((feedHash) => {
+        const mint = mintMap.get(feedHash)?.toBase58();
+        if (!mint) {
+          console.error("Error:", `Mint not found for feedHash ${feedHash}`);
+          brokenFeeds.push(feedHash);
+        }
+        return mint;
+      })
+      .filter((mint): mint is string => mint !== undefined);
+
+    const response = await fetchMultiPrice(tokens);
+
+    const priceData = response.data;
+
+    const finalPayload: CrossbarSimulatePayload = feedHashes.map((feedHash) => {
+      const tokenAddress = mintMap.get(feedHash)!.toBase58();
+      const price = priceData[tokenAddress];
+      return {
+        feedHash,
+        results: [price.value],
+      };
+    });
+
+    return { payload: finalPayload, brokenFeeds };
+  } catch (error) {
+    console.error("Error:", error);
+    return { payload: [], brokenFeeds: feedHashes };
+  }
+}
+
+async function fetchCrossbarPrices(
+  feedHashes: string[],
+  endpoint: string,
+  username?: string,
+  bearer?: string
+): Promise<{ payload: CrossbarSimulatePayload; brokenFeeds: string[] }> {
   const controller = new AbortController();
   const timeoutId = setTimeout(() => {
     controller.abort();
-  }, 5000);
+  }, 8000);
+
+  const isAuth = username && bearer;
+
+  const basicAuth = isAuth ? Buffer.from(`${username}:${bearer}`).toString("base64") : undefined;
 
   try {
     const feedHashesString = feedHashes.join(",");
-    const response = await fetch(`${SWITCHBOARD_CROSSSBAR_API}/simulate/${feedHashesString}`, {
+    const response = await fetch(`${endpoint}/simulate/${feedHashesString}`, {
       headers: {
+        Authorization: basicAuth ? `Basic ${basicAuth}` : "",
```

Comment:
Basic ${basicAuth} -> should this have a space inbetween? Also, doesnt it cause any issues if we have an empty Authorization string?

Reply:
yeah the space is needed, the public crossbar api doesn't need authorization so this header is ignored